session-helper/background-monitor.c

```diff
@@ -0,0 +1,426 @@
+}
+
+static void
+forbid_app (NotifyNotification *notification,
+            char               *action,
+            gpointer            user_data)
+{
+  FlatpakInstance *instance = user_data;
+
+  g_debug ("Forbid app %s to run in background", flatpak_instance_get_app (instance));
+
+  set_permission (flatpak_instance_get_app (instance), NO);
+  kill_instance (instance);
+
+  remove_notification (notification);
+}
+
+static void
+ignore_app (NotifyNotification *notification,
+            char               *action,
+            gpointer            user_data)
+{
+  FlatpakInstance *instance = user_data;
+
+  g_debug ("Let app %s run in background", flatpak_instance_get_app (instance));
+
+  remove_notification (notification);
+}
+
+static void
+stop_app (NotifyNotification *notification,
+          char               *action,
+          gpointer            user_data)
+{
+  FlatpakInstance *instance = user_data;
+
+  g_debug ("Stop app %s", flatpak_instance_get_app (instance));
+
+  kill_instance (instance);
+
+  remove_notification (notification);
+}
+
+static void
+send_notification (FlatpakInstance *instance,
+                   Permission       permission)
+{
+  g_autofree char *name = flatpak_instance_get_display_name (instance);
+  const char *summary;
+  g_autofree char *body = NULL;
+  NotifyNotification *notification;
+
+  summary = _("Background activity");
+  body = g_strdup_printf (_("%s is running in the background."), name);
+  notification = notify_notification_new (summary, body, NULL);
+
+  if (permission == UNSET)
+    {
+      notify_notification_add_action (notification, "allow", _("Allow"), allow_app, g_object_ref (instance), g_object_unref);
+      notify_notification_add_action (notification, "forbid", _("Forbid"), forbid_app, g_object_ref (instance), g_object_unref);
+    }
+  else if (permission == ASK)
+    {
+      notify_notification_add_action (notification, "ignore", _("Ignore"), ignore_app, g_object_ref (instance), g_object_unref);
+      notify_notification_add_action (notification, "stop", _("Stop"), stop_app, g_object_ref (instance), g_object_unref);
+    }
+
+  add_notification_for_instance (instance, notification);
+  g_object_unref (notification);
+}
+
+static void
+thread_func (GTask *task,
+             gpointer source_object,
+             gpointer task_data,
+             GCancellable *cancellable)
+{
+  g_autoptr(GVariant) perms = NULL;
+  g_autoptr(GPtrArray) apps = NULL;
+  int i;
+
+  g_debug ("Checking background permissions");
+
+  perms = get_permissions ();
+
+  apps = flatpak_instance_get_all ();
+
+  remove_outdated_notifications (apps);
+
+  for (i = 0; i < apps->len; i++)
+    {
+      FlatpakInstance *instance = g_ptr_array_index (apps, i);
+      const char *app_id = flatpak_instance_get_app (instance);
+      Permission permission;
+
+      if (!flatpak_instance_is_running (instance))
+        continue;
+
+      if (flatpak_instance_has_window (instance))
+        continue;
```

Comment:
This is a bity racy. I mean, clearly its not 100% due to the timeout. But, it may very well be that an app *just* started and hit the check, and hasn't mapped a window yet. I think we should poll with some regularly, but if there is a "new" instance we should wait for a few seconds more and then check again for a window before asking about it.

Reply:
good point